src/aliceVision/image/curves.cpp

```diff
@@ -0,0 +1,250 @@
+#include <cmath>
+
+#include "curves.hpp"
+
+
+namespace alicevision {
+    namespace image {
+
+        // Tone curve according to Adobe's reference implementation
+        // values in 0xffff space
+        void AdobeToneCurve::Apply(float& ir, float& ig, float& ib) const
+        {
+
+            float r = std::max<float>(0.0, std::min<float>(65535.0, ir));
+            float g = std::max<float>(0.0, std::min<float>(65535.0, ig));
+            float b = std::max<float>(0.0, std::min<float>(65535.0, ib));
+
+            if (r >= g) {
+                if (g > b) {
+                    RGBTone(r, g, b);     // Case 1: r >= g >  b
+                }
+                else if (b > r) {
+                    RGBTone(b, r, g);     // Case 2: b >  r >= g
+                }
+                else if (b > g) {
+                    RGBTone(r, b, g);     // Case 3: r >= b >  g
+                }
+                else {                           // Case 4: r == g == b
+                    r = lutToneCurve[r];
+                    g = lutToneCurve[g];
+                    b = g;
+                }
+            }
+            else {
+                if (r >= b) {
+                    RGBTone(g, r, b);     // Case 5: g >  r >= b
+                }
+                else if (b > g) {
+                    RGBTone(b, g, r);     // Case 6: b >  g >  r
+                }
+                else {
+                    RGBTone(g, b, r);     // Case 7: g >= b >  r
+                }
+            }
+
+            //setUnlessOOG(ir, ig, ib, r, g, b);
+            if (!(ir < 0.0 || ir > 65535.0) || !(ig < 0.0 || ig > 65535.0) || !(ib < 0.0 || ib > 65535.0)) {
+                ir = r;
+                ig = g;
+                ib = b;
+            }
+
+        }
+
+
+        inline void AdobeToneCurve::RGBTone(float& maxval, float& medval, float& minval) const
+        {
+            float minvalold = minval, medvalold = medval, maxvalold = maxval;
+
+            maxval = lutToneCurve[maxvalold];
+            minval = lutToneCurve[minvalold];
+            medval = minval + ((maxval - minval) * (medvalold - minvalold) / (maxvalold - minvalold));
+        }
+
+        void ToneCurve::Reset()
+        {
+            lutToneCurve.reset();
+        }
+
+        // Fill a LUT with X/Y, ranged 0xffff
+        void ToneCurve::Set(const Curve& pCurve, float gamma)
+        {
+            lutToneCurve(65536);
+
+            if (gamma <= 0.0 || gamma == 1.) {
+                for (int i = 0; i < 65536; i++) {
+                    lutToneCurve[i] = (float)pCurve.getVal(float(i) / 65535.f) * 65535.f;
+                }
+
+            }
+        }
+
+
+        Curve::Curve() : N(0), ppn(0), x(nullptr), y(nullptr), mc(0.0), mfc(0.0), msc(0.0), mhc(0.0), hashSize(1000 /* has to be initialized to the maximum value */), ypp(nullptr), x1(0.0), y1(0.0), x2(0.0), y2(0.0), x3(0.0), y3(0.0), firstPointIncluded(false), increment(0.0), nbr_points(0) {}
```

Comment:
maybe initialize the field directly in the hpp and keep this constructor to default (redeability)
```c++
Curve() = default;
```